### PR TITLE
Fix timing issues in RunOnly tests

### PR DESCRIPTION
**Pull request by wtgodbe**
Attempts to resolve https://github.com/dotnet/coreclr/issues/11503 and https://github.com/dotnet/coreclr/issues/11502

CC @gkhanna79 @RussKeldorph

### CODE REVIEW

**Comment by wtgodbe** [comment]
@dotnet-bot test test Ubuntu Release pri1r2r

**Comment by wtgodbe** [comment]
Also @ramarag PTAL

File: tests/src/CLRTest.Execute.Bash.targets [hunk 1]
```diff
@@ -356,8 +359,20 @@ ReleaseLock()
         return 3
     fi
 }
+
+CheckForCrossgenMarker()
+{
+    echo "in CheckForCrossgenMarker"
+    if [ ! -f "$MarkerFile" ]%3B then
+        sleep 1
+        CheckForCrossgenMarker
```
Comment (RussKeldorph) [inline tests/src/CLRTest.Execute.Bash.targets]: Unbounded recursion seems unnecessary.  Wouldn't a simple `for` or `while` loop be better.

**Comment by RussKeldorph** [comment]
@wtgodbe Can you explain how this works?  For `CheckForCrossgenMarker` exit, I expect something to delete `MarkerFile`, but I don't see that happening?  Do I misunderstand?

Also, I'm not following how this avoids a race condition.  It seems like the simplest thing would be to expand the `TakeLock` logic to just cover both crossgen'ing and execution of the test.  Maybe `CLRTestBa*PostCommands` would be useful for that.  Not pretty, but perhaps better than now.

Also, how does this address #11502?  Seems like you're still modifying the test binaries.

**Comment by ramarag** [comment]
if this is only meant for Runonly tests, i would suggest getting rid of this category altogether, they are causing more grief than benefits. Last i saw there were only a few remaining.

**Comment by RussKeldorph** [comment]
Ah, I misunderstood at least one thing: it looks like each test waits until MarkerFile is created and nothing needs to delete it.  In that case, shouldn't only the test that did the crossgen create MarkerFile?  I'm not following how that's an exclusive task.

**Comment by RussKeldorph** [comment]
@ramarag
> Last i saw there were only a few remaining.

```
C:\git\coreclr\tests\src>findstr -spn CLRTestProjectToRun *proj|findstr -v GCSimulator|wc -l
         52

C:\git\coreclr\tests\src>findstr -spn CLRTestProjectToRun *proj|findstr GCSimulator|wc -l
        430
```

:-/

**Comment by wtgodbe** [comment]
@RussKeldorph it turns out that only the Build-Only tests attempt to crossgen the .exe file, while the run-only tests just attempt to Corerun whatever .exe is present - so only 1 project will create the marer file, then all of the run-only's wait for it to exist before executing

**Comment by wtgodbe** [comment]
I had thought of surrounding the call to corerun with TakeLock, but then we lose parallelism on executing these tests

**Comment by wtgodbe** [comment]
I'm also not sure it's possible to not modify test binaries at all - at the very least, we need to replace the original .exe with the crossgen'd exe at some point, meaning that during that operation there will be a period where there is no .exe present in the test output dir. This attempts to circumvent that problem by ensuring that nobody tries to execute the test .exe during or before Crossgen

**Comment by ramarag** [comment]
@RussKeldorph

I believe the gcsimulator tests do not use the normal test infra to run them, @swgillespie am i correct ?

**Comment by wtgodbe** [comment]
I agree we should get rid of those tests if possible, and can file an issue against myself to do so for 2.1, but we need to get these tests passing for 2.0.0, which means getting this change in today.

**Comment by wtgodbe** [comment]
@dotnet-bot test this please

**Comment by wtgodbe** [comment]
@dotnet-bot test test Ubuntu Release pri1r2r

**Comment by wtgodbe** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build

**Comment by RussKeldorph** [comment]
> it turns out that only the Build-Only tests attempt to crossgen the .exe file

In that case, I don't know why the TakeLock construct is necessary.  By "Build-Only" Are you referring to `<CLRTestKind>`?  I guess with that you could use that to ensure no race conditions, but it seems fragile since you're relying on the tests themselves to to conform to a vague contract.  For example, the tests under `baseservices\threading\paramthreadstart` use `RunOnly` but not `BuildOnly`.  Would those need to be fixed to conform to your assumption, or is `BuildOnly` implied somewhere?  (Sorry for asking so many questions; I haven't had a chance to dig through every line of this.)

> I'm also not sure it's possible to not modify test binaries at all - at the very least, we need to replace the original .exe with the crossgen'd exe at some point

We could do all the writing to a temp folder.  I suspect the vast majority of tests don't have a dependency on their own path location.  For the tests that do, we should give them a way to refer to their original location or copy necessary dependencies to their "working folder."  At any rate, I'd prefer not to close #11502 this this PR.  That should only be closed if we really stop writing to the tests folder or we decide it's not worth fixing.

**Comment by RussKeldorph** [comment]
> is BuildOnly implied somewhere?

I suspect this this mechanism works as long as, among a group of tests that share a binary, there is at most one that is *not* `RunOnly`, so a single `BuildOnly` or a single Build-and-Run is fine.  If that's the explanation, then Looks Ok To Me.  Not great, but ok.

**Comment by swgillespie** [comment]
@ramarag 

> I believe the gcsimulator tests do not use the normal test infra to run them, @swgillespie am i correct ?

GCSimulator tests use the normal test infra, it skips all tests not marked with the GCSimulator test attribute.

**Comment by wtgodbe** [comment]
@dotnet-bot test this please

**Comment by wtgodbe** [comment]
@dotnet-bot test test Ubuntu Release pri1r2r

**Pull request closed without merging**